probe/endpoint/reporter.go

```diff
@@ -79,140 +87,148 @@ func (r *Reporter) Report() (report.Report, error) {
+	{
+		extraNodeInfo := report.MakeNode().WithLatests(map[string]string{
+			Conntracked: "true",
+		})
+		r.flowWalker.walkFlows(func(f flow) {
+			var (
+				fromAddr, fromPort = f.Original.Layer3.SrcIP, uint16(f.Original.Layer4.SrcPort)
+				toAddr, toPort     = f.Original.Layer3.DstIP, uint16(f.Original.Layer4.DstPort)
+				tuple              = []string{
+					fmt.Sprintf("%s:%d", fromAddr, fromPort),
+					fmt.Sprintf("%s:%d", toAddr, toPort),
+				}
+			)
+			sort.Strings(tuple)
+			seenTuples[strings.Join(tuple, " ")] = fourTuple{fromAddr, toAddr, fromPort, toPort}
+			r.addConnection(&rpt, fromAddr, toAddr, fromPort, toPort, &extraNodeInfo, &extraNodeInfo)
+		})
+	}
 
 	{
 		conns, err := r.scanner.Connections(r.includeProcesses)
 		if err != nil {
 			return rpt, err
 		}
-		commonNodeInfo := report.MakeNode().WithLatests(map[string]string{
+		toNodeInfo := report.MakeNode().WithLatests(map[string]string{
 			Procspied: "true",
 		})
 		for conn := conns.Next(); conn != nil; conn = conns.Next() {
 			var (
-				localPort  = conn.LocalPort
-				remotePort = conn.RemotePort
-				localAddr  = conn.LocalAddress.String()
-				remoteAddr = conn.RemoteAddress.String()
+				fromAddr, fromPort = conn.LocalAddress.String(), conn.LocalPort
+				toAddr, toPort     = conn.RemoteAddress.String(), conn.RemotePort
+				tuple              = []string{
+					fmt.Sprintf("%s:%d", fromAddr, fromPort),
+					fmt.Sprintf("%s:%d", toAddr, toPort),
+				}
 			)
-			extraNodeInfo := commonNodeInfo.Copy()
+			fromNodeInfo := toNodeInfo.Copy()
 			if conn.Proc.PID > 0 {
-				extraNodeInfo = extraNodeInfo.WithLatests(map[string]string{
+				fromNodeInfo = fromNodeInfo.WithLatests(map[string]string{
 					process.PID:       strconv.FormatUint(uint64(conn.Proc.PID), 10),
 					report.HostNodeID: hostNodeID,
 				})
 			}
-			r.addConnection(&rpt, localAddr, remoteAddr, localPort, remotePort, &extraNodeInfo, &commonNodeInfo)
+			sort.Strings(tuple)
+
+			// If we've already seen this connection, we should know the direction
+			// (or have already figured it out), so we normalize and use the
+			// canonical direction. Otherwise, we can use a port-heuristic to guess
+			// the direction.
+			if c, ok := seenTuples[strings.Join(tuple, " ")]; ok {
+				if fromAddr != c.fromAddr || fromPort != c.fromPort {
+					fromAddr, fromPort, toAddr, toPort = c.fromAddr, c.fromPort, c.toAddr, c.toPort
+					toNodeInfo, fromNodeInfo = fromNodeInfo, toNodeInfo
+				}
+			} else if int(fromPort) < int(toPort) {
+				fromAddr, fromPort, toAddr, toPort = toAddr, toPort, fromAddr, fromPort
+				toNodeInfo, fromNodeInfo = fromNodeInfo, toNodeInfo
+			}
+			seenTuples[strings.Join(tuple, " ")] = fourTuple{fromAddr, toAddr, fromPort, toPort}
+			// TODO(paulbellamy): Won't adding this from conntrack, and adding the
+			// edge again here result in double-counting? Also, if we've seen the
+			// same conn twice from proc (if both ends are on the same host).
```

Comment:
Yes this will double count the MaxConnCountTCP, but as we don't show that anywhere, I think we should remove it.